File: LR4/LR4/voting_system.py
class VotingSystem:
    def __init__(self, candidates, voters):
        self.num_candidates = candidates
        self.num_voters = voters
        self.participants = []
        self.ballots = []
    
    def input_candidates(self):
        print("Введите имена кандидатов:")
        for i in range(self.num_candidates):
            candidate = input(f"Кандидат {i+1}: ").strip()
            self.participants.append(candidate)
    
    def input_votes(self):
        print("Введите предпочтения избирателей (от лучшего к худшему):")
        for i in range(self.num_voters):
            print(f"Избиратель {i+1}:")
            ballot = []
            for j in range(self.num_candidates):
                vote = input(f"  Позиция {j+1}: ").strip()
                ballot.append(vote)
            self.ballots.append(ballot)
    
    def calculate_borda_winner(self):
        points = [0] * self.num_candidates
        
        for ballot in self.ballots:
            for position in range(self.num_candidates):
                candidate = ballot[position]
                try:
                    candidate_index = self.participants.index(candidate)
                    points[candidate_index] += (self.num_candidates - 1 - position)
                except ValueError:
                    print(f"Внимание: кандидат '{candidate}' не найден в списке")
        
        max_points = max(points)
        winner_index = points.index(max_points)
        
        print("\nМетод Борда:")
        for i in range(self.num_candidates):
            print(f"{self.participants[i]}: {points[i]} баллов")
        print(f"Победитель: {self.participants[winner_index]}")
        
        return self.participants[winner_index]
    
    def calculate_condorcet_winner(self):
        wins = [[0 for _ in range(self.num_candidates)] for _ in range(self.num_candidates)]
        
        for i in range(self.num_candidates):
            for j in range(self.num_candidates):
                if i == j:
                    continue
                
                wins_count = 0
                for ballot in self.ballots:
                    try:
                        pos_i = ballot.index(self.participants[i])
                        pos_j = ballot.index(self.participants[j])
                        
                        if pos_i < pos_j:
                            wins_count += 1
                    except ValueError:
                        print(f"Внимание: кандидат не найден в бюллетене")
                
                if wins_count > self.num_voters // 2:
                    wins[i][j] = 1
        
        for i in range(self.num_candidates):
            is_condorcet_winner = True
            for j in range(self.num_candidates):
                if i != j and wins[i][j] == 0:
                    is_condorcet_winner = False
                    break
            
            if is_condorcet_winner:
                print(f"\nМетод Кондорсе:\nПобедитель: {self.participants[i]}")
                return self.participants[i]
        
        print("\nМетод Кондорсе:\nПобедителя нет")
        return ""
    
    def run_election(self):
        self.input_candidates()
        self.input_votes()
        
        borda_winner = self.calculate_borda_winner()
        condorcet_winner = self.calculate_condorcet_winner()
        
        if condorcet_winner and borda_winner != condorcet_winner:
            print("\nМетоды дали разные результаты!")

def show_test_cases():
    print("\nПримеры тестирования:")
    print("1. Единогласие: все голосуют A B C → A побеждает везде")
    print("2. Парадокс: A B C, B C A, C A B → нет победителя Кондорсе")
    print("3. Разногласия: возможны разные победители в методах")

def show_method_analysis():
    print("\nСравнение методов:")
    print("Борда: всегда есть победитель, учитывает все предпочтения")
    print("Кондорсе: наиболее справедлив, но может не найти победителя")
    print("Разные результаты при циклических предпочтениях")

def main():
    print("Система голосования (методы Борда и Кондорсе)")
    print("1 - Голосование, 2 - Тесты, 3 - Анализ: ", end="")
    
    try:
        choice = int(input())
        
        if choice == 1:
            candidates = int(input("Количество кандидатов: "))
            voters = int(input("Количество избирателей: "))
            
            election = VotingSystem(candidates, voters)
            election.run_election()
        elif choice == 2:
            show_test_cases()
        elif choice == 3:
            show_method_analysis()
        else:
            print("Неверный выбор")
    except ValueError:
        print("Ошибка: введите число")

if __name__ == "__main__":
    main()

File: LR4/LR4/voting_system.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <numeric>
#include <iomanip>

using namespace std;

class VotingSystem {
private:
    vector<string> participants;
    vector<vector<string>> ballots;
    int numCandidates;
    int numVoters;

public:
    VotingSystem(int candidates, int voters) : numCandidates(candidates), numVoters(voters) {
        participants.resize(candidates);
        ballots.resize(voters);
    }

    void inputCandidates() {
        cout << "Введите имена кандидатов:\n";
        for (int i = 0; i < numCandidates; i++) {
            cin >> participants[i];
        }
    }

    void inputVotes() {
        cout << "Введите предпочтения избирателей (от лучшего к худшему):\n";
        for (int i = 0; i < numVoters; i++) {
            ballots[i].resize(numCandidates);
            for (int j = 0; j < numCandidates; j++) {
                cin >> ballots[i][j];
            }
        }
    }

    string calculateBordaWinner() {
        vector<int> points(numCandidates, 0);

        for (const auto& ballot : ballots) {
            for (int position = 0; position < numCandidates; position++) {
                string candidate = ballot[position];
                auto it = find(participants.begin(), participants.end(), candidate);
                int candidateIndex = distance(participants.begin(), it);
                points[candidateIndex] += (numCandidates - 1 - position);
            }
        }

        int maxPoints = *max_element(points.begin(), points.end());
        int winnerIndex = distance(points.begin(),
            find(points.begin(), points.end(), maxPoints));

        cout << "\nМетод Борда:\n";
        for (int i = 0; i < numCandidates; i++) {
            cout << participants[i] << ": " << points[i] << " баллов\n";
        }
        cout << "Победитель: " << participants[winnerIndex] << "\n";

        return participants[winnerIndex];
    }

    string calculateCondorcetWinner() {
        vector<vector<int>> wins(numCandidates, vector<int>(numCandidates, 0));

        for (int i = 0; i < numCandidates; i++) {
            for (int j = 0; j < numCandidates; j++) {
                if (i == j) continue;

                int winsCount = 0;
                for (const auto& ballot : ballots) {
                    int posI = distance(ballot.begin(),
                        find(ballot.begin(), ballot.end(), participants[i]));
                    int posJ = distance(ballot.begin(),
                        find(ballot.begin(), ballot.end(), participants[j]));

                    if (posI < posJ) {
                        winsCount++;
                    }
                }

                if (winsCount > numVoters / 2) {
                    wins[i][j] = 1;
                }
            }
        }

        for (int i = 0; i < numCandidates; i++) {
            bool isCondorcetWinner = true;
            for (int j = 0; j < numCandidates; j++) {
                if (i != j && wins[i][j] == 0) {
                    isCondorcetWinner = false;
                }
            }

            if (isCondorcetWinner) {
                cout << "\nМетод Кондорсе:\nПобедитель: " << participants[i] << "\n";
                return participants[i];
            }
        }

        cout << "\nМетод Кондорсе:\nПобедителя нет\n";
        return "";
    }

    void runElection() {
        inputCandidates();
        inputVotes();

        string bordaWinner = calculateBordaWinner();
        string condorcetWinner = calculateCondorcetWinner();

        if (!condorcetWinner.empty() && bordaWinner != condorcetWinner) {
            cout << "\nМетоды дали разные результаты!\n";
        }
    }
};

void showTestCases() {
    cout << "\nПримеры тестирования:\n";
    cout << "1. Единогласие: все голосуют A B C → A побеждает везде\n";
    cout << "2. Парадокс: A B C, B C A, C A B → нет победителя Кондорсе\n";
    cout << "3. Разногласия: возможны разные победители в методах\n";
}

void showMethodAnalysis() {
    cout << "\nСравнение методов:\n";
    cout << "Борда: всегда есть победитель, учитывает все предпочтения\n";
    cout << "Кондорсе: наиболее справедлив, но может не найти победителя\n";
    cout << "Разные результаты при циклических предпочтениях\n";
}

int main() {
    setlocale(LC_ALL, "ru");
    cout << "Система голосования (методы Борда и Кондорсе)\n";
    cout << "1 - Голосование, 2 - Тесты, 3 - Анализ: ";

    int choice;
    cin >> choice;

    switch (choice) {
    case 1: {
        int candidates, voters;
        cout << "Количество кандидатов: ";
        cin >> candidates;
        cout << "Количество избирателей: ";
        cin >> voters;

        VotingSystem election(candidates, voters);
        election.runElection();
        break;
    }
    case 2:
        showTestCases();
        break;
    case 3:
        showMethodAnalysis();
        break;
    default:
        cout << "Неверный выбор\n";
    }

    return 0;
}